src/FEValues/FunctionValues.jl

```diff
@@ -0,0 +1,184 @@
+sdim_from_gradtype(::Type{<:SVector{sdim}}) where sdim = sdim
+sdim_from_gradtype(::Type{<:SMatrix{<:Any,sdim}}) where sdim = sdim
+
+# For performance, these must be fully inferrable for the compiler.
+# args: valname (:CellValues or :FaceValues), shape_gradient_type, eltype(x)
+function check_reinit_sdim_consistency(valname, gradtype, ::Type{<:Vec{sdim}}) where {sdim}
+    check_reinit_sdim_consistency(valname, Val(sdim_from_gradtype(gradtype)), Val(sdim))
+end
+check_reinit_sdim_consistency(_, ::Val{sdim}, ::Val{sdim}) where sdim = nothing
+function check_reinit_sdim_consistency(valname, ::Val{sdim_val}, ::Val{sdim_x}) where {sdim_val, sdim_x}
+    error("""The $valname and coordinates have different spatial dimensions, $sdim_val and $sdim_x.
+        Could it be that you forgot to vectorize the geometric interpolation when constructing the $valname?""")
+end
+
+# Mapping types 
+struct IdentityMapping end 
+struct CovariantPiolaMapping end 
+struct ContravariantPiolaMapping end 
+# Not yet implemented:
+# struct DoubleCovariantPiolaMapping end 
+# struct DoubleContravariantPiolaMapping end 
+
+get_mapping_type(fv::FunctionValues) = get_mapping_type(fv.ip)
+
+"""
+    requires_hessian(mapping)
+
+Does the `mapping` type require the hessian, d²M/dx², 
+to map the function values and gradients from the reference cell 
+to the real cell geometry?
+"""
+requires_hessian(::IdentityMapping) = false
+requires_hessian(::ContravariantPiolaMapping) = true
+requires_hessian(::CovariantPiolaMapping) = true
+
+# Support for embedded elements
+calculate_Jinv(J::Tensor{2}) = inv(J)
+calculate_Jinv(J::SMatrix) = pinv(J)
+
+# Hotfix to get the dots right for embedded elements until mixed tensors are merged.
+# Scalar/Vector interpolations with sdim == rdim (== vdim)
+@inline dothelper(A, B) = A ⋅ B
+# Vector interpolations with sdim == rdim != vdim
+@inline dothelper(A::SMatrix{vdim, dim}, B::Tensor{2, dim}) where {vdim, dim} = A * SMatrix{dim, dim}(B)
+# Scalar interpolations with sdim > rdim
+@inline dothelper(A::SVector{rdim}, B::SMatrix{rdim, sdim}) where {rdim, sdim} = B' * A
+# Vector interpolations with sdim > rdim
+@inline dothelper(B::SMatrix{vdim, rdim}, A::SMatrix{rdim, sdim}) where {vdim, rdim, sdim} = B * A
+
+@inline function apply_mapping!(funvals::FunctionValues, args...)
+    return apply_mapping!(funvals, get_mapping_type(funvals), args...)
+end
+
+@inline function apply_mapping!(funvals::FunctionValues, ::IdentityMapping, q_point::Int, mapping_values, args...)
+    Jinv = calculate_Jinv(getjacobian(mapping_values))
+    @inbounds for j in 1:getnbasefunctions(funvals)
+        #funvals.dNdx[j, q_point] = funvals.dNdξ[j, q_point] ⋅ Jinv # TODO via Tensors.jl
+        funvals.dNdx[j, q_point] = dothelper(funvals.dNdξ[j, q_point], Jinv)
+    end
+    return nothing
+end
+
+@inline function apply_mapping!(funvals::FunctionValues, ::CovariantPiolaMapping, q_point::Int, mapping_values, cell)
```

Comment:
Perhaps comment out for now? I assume this isn't tested until #798?

Reply:
Yes, wasn't sure how much to leave out since this is an important part of the motivation.
Will comment out these as well as the then unused mapping types!
And add a reference in the comment to PR798.